### PR TITLE
Issue.swift41

### PR DESCRIPTION
**Pull request by ShihabMehboob**
Resolves issues with Swift 4.1:
- Removes compilation warnings by replacing `flatMap` with `compactMap` when compiling on Swift >=4.1
- Fixes a bug where `Coder.getFieldName(codingPath:)` returns an erroneous verbose description of the CodingKeys rather than the simple String value.

### CODE REVIEW

**Comment by ianpartridge** [comment]
Needs rebasing.

File: Sources/KituraContracts/CodableQuery/Coder.swift [hunk 1]
```diff
@@ -32,6 +32,10 @@ public class Coder {
 
     /// Helper method to extract the field name from a CodingKey array
     public static func getFieldName(from codingPath: [CodingKey]) -> String {
-        return codingPath.flatMap({"\($0)"}).joined(separator: ".")
+		#if swift(>=4.1) || os(Linux)
```
Comment (ianpartridge) [inline Sources/KituraContracts/CodableQuery/Coder.swift]: This doesn't make sense - it changes behaviour for Swift 4 on Linux?

File: Sources/KituraContracts/CodableQuery/QueryEncoder.swift [hunk 2]
```diff
@@ -46,6 +46,18 @@ public class QueryEncoder: Coder, Encoder {
         return "?" + String(desc.dropFirst())
     }
 
+    /// Encodes an Encodable object to a URLQueryItem array
```
Comment (ianpartridge) [inline Sources/KituraContracts/CodableQuery/QueryEncoder.swift]: What's this code?

**Review by ianpartridge — CHANGES_REQUESTED** [review]

File: Sources/KituraContracts/CodableQuery/Coder.swift (hunk 1 above)
Comment (ShihabMehboob) [reply Sources/KituraContracts/CodableQuery/Coder.swift]: Amended this now, there were failing Linux travis builds and thought it was to do with that. It wasn't. Resolved that now.

File: Sources/KituraContracts/CodableQuery/QueryEncoder.swift (hunk 2 above)
Comment (ShihabMehboob) [reply Sources/KituraContracts/CodableQuery/QueryEncoder.swift]: That's not something I'd added, it's already in master. Not sure why that got shown as changed.

**Comment by djones6** [comment]
@ianpartridge looks like @ShihabMehboob has addressed your review comments.

**Review by ianpartridge — APPROVED** [review]
Looks good now.

**Comment by djones6** [comment]
@ShihabMehboob FYI, I updated the opening comment of this PR to be more descriptive.

**Pull request merged**